crates/sui-types/src/passkey_authenticator.rs

```diff
@@ -0,0 +1,191 @@
+    digests::ZKLoginInputsDigest,
+    error::{SuiError, SuiResult},
+    signature::{AuthenticatorTrait, VerifyParams},
+};
+use fastcrypto::hash::{HashFunction, Sha256};
+use fastcrypto::rsa::{Base64UrlUnpadded, Encoding};
+use fastcrypto::secp256r1::{Secp256r1PublicKey, Secp256r1Signature};
+use fastcrypto::traits::VerifyingKey;
+use fastcrypto::{error::FastCryptoError, traits::ToFromBytes};
+use once_cell::sync::OnceCell;
+use passkey::types::webauthn::CollectedClientData;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use shared_crypto::intent::IntentMessage;
+use std::hash::Hash;
+use std::hash::Hasher;
+use std::sync::Arc;
+
+#[cfg(test)]
+#[path = "unit_tests/passkey_authenticator_test.rs"]
+mod passkey_authenticator_test;
+
+/// An zk login authenticator with all the necessary fields.
+#[derive(Debug, Clone, JsonSchema, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct PasskeyAuthenticator {
+    authenticator_data: Vec<u8>, // should size limit be enforced?
+    client_data: Vec<u8>,        // should size limit be enforced?
+    user_signature: Signature,
+    #[serde(skip)]
+    pub bytes: OnceCell<Vec<u8>>,
+}
+
+impl PasskeyAuthenticator {
+    // test only
+    pub fn new(
+        authenticator_data: Vec<u8>,
+        client_data: Vec<u8>,
+        user_signature: Signature,
+    ) -> Self {
+        Self {
+            authenticator_data,
+            client_data,
+            user_signature,
+            bytes: OnceCell::new(),
+        }
+    }
+
+    pub fn get_pk(&self) -> SuiResult<PublicKey> {
+        PublicKey::try_from_bytes(
+            SignatureScheme::Secp256r1,
+            self.user_signature.public_key_bytes(),
+        )
+        .map_err(|_| SuiError::InvalidAuthenticator)
+    }
+}
+
+/// Necessary trait for [struct SenderSignedData].
+impl PartialEq for PasskeyAuthenticator {
+    fn eq(&self, other: &Self) -> bool {
+        self.as_ref() == other.as_ref()
+    }
+}
+
+/// Necessary trait for [struct SenderSignedData].
+impl Eq for PasskeyAuthenticator {}
+
+/// Necessary trait for [struct SenderSignedData].
+impl Hash for PasskeyAuthenticator {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.as_ref().hash(state);
+    }
+}
+
+impl AuthenticatorTrait for PasskeyAuthenticator {
+    fn verify_user_authenticator_epoch(
+        &self,
+        _epoch: EpochId,
+        _max_epoch_upper_bound_delta: Option<u64>,
+    ) -> SuiResult {
+        Ok(())
+    }
+
+    /// Verify an intent message of a transaction with an zk login authenticator.
+    fn verify_claims<T>(
+        &self,
+        intent_msg: &IntentMessage<T>,
+        author: SuiAddress,
+        _aux_verify_data: &VerifyParams,
+        _zklogin_inputs_cache: Arc<VerifiedDigestCache<ZKLoginInputsDigest>>,
+    ) -> SuiResult
+    where
+        T: Serialize,
+    {
+        let client_data_json: CollectedClientData = serde_json::from_slice(&self.client_data)
```

Comment:
so client data needs to be this `CollectedClientData` type but do we want to/need to enforce any other checks on the structure of this data?

Reply:
In particular we likely should force checking some of the fields. For reference here are what they are:

```
pub struct CollectedClientData {
    pub ty: [ClientDataType](https://docs.rs/passkey-types/0.2.0/passkey_types/webauthn/enum.ClientDataType.html),
    pub challenge: [String](https://doc.rust-lang.org/nightly/alloc/string/struct.String.html),
    pub origin: [String](https://doc.rust-lang.org/nightly/alloc/string/struct.String.html),
    pub cross_origin: [Option](https://doc.rust-lang.org/nightly/core/option/enum.Option.html)<[bool](https://doc.rust-lang.org/nightly/std/primitive.bool.html)>,
    pub unknown_keys: [IndexMap](https://docs.rs/indexmap/2.1.0/x86_64-unknown-linux-gnu/indexmap/map/struct.IndexMap.html)<[String](https://doc.rust-lang.org/nightly/alloc/string/struct.String.html), [Value](https://docs.rs/serde_json/1.0.108/x86_64-unknown-linux-gnu/serde_json/value/enum.Value.html)>,
}
```

Reply:
maybe we can check ClientDataType is indeed webauthn.get. perhaps check on origin for some requirements that we can discuss.